### PR TITLE
Management command for creating bulk orders

### PR DESCRIPTION
**Pull request by ahsan-ul-haq**
Bulk orders are needed to create for testing migrations on
huge date
LEARNER-3075

### CODE REVIEW

File: ecommerce/core/management/commands/generate_bulk_orders.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+from __future__ import unicode_literals
```
Comment (clintonb) [inline ecommerce/core/management/commands/generate_bulk_orders.py]: Move this to ecommerce/extentions/order/management/commands/create_fake_orders.py.

We want to keep order-related code together. Also, we want to make it clear that this data is fake and should not be used in production.

File: ecommerce/core/management/commands/generate_bulk_orders.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    basket.strategy = Default()
+    basket.add_product(product)
+    return basket
+
+
+class Command(BaseCommand):
+    help = 'Clean history data'
+
+    def add_arguments(self, parser):
+        parser.add_argument('--orders',
```
Comment (clintonb) [inline ecommerce/core/management/commands/generate_bulk_orders.py]: Let's call this count so that its purpose is obvious.

File: ecommerce/core/management/commands/generate_bulk_orders.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                            action='store',
+                            dest='orders',
+                            type=int,
+                            required=True,
+                            help='Number of orders to be generated as bulk order.')
```
Comment (clintonb) [inline ecommerce/core/management/commands/generate_bulk_orders.py]: The term "bulk order" implies something different for me. I would rephrase this to: Number of orders to create.

File: ecommerce/core/management/commands/generate_bulk_orders.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        parser.add_argument('--product_id',
```
Comment (clintonb) [inline ecommerce/core/management/commands/generate_bulk_orders.py]: Use SKU as that is how we generally present products to end users/API clients. Also, the corresponding `StockRecord` will keep track of the price and currency being used.

File: ecommerce/core/management/commands/generate_bulk_orders.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+        if product_id:
+            try:
+                product = Product.objects.get(id=product_id)
+            except Product.DoesNotExist:
+                msg = 'No product for product_id {} exists.'.format(product_id)
+                logger.exception(msg)
+                raise CommandError(msg)
+
+            if not product.is_seat_product:
```
Comment (clintonb) [inline ecommerce/core/management/commands/generate_bulk_orders.py]: Why does the product class matter? I see no reason to limit this to seat products.

File: ecommerce/core/management/commands/generate_bulk_orders.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                            action='store',
+                            dest='product_id',
+                            type=int,
+                            help='Product Id for which orders are to be generated.')
+
+    def handle(self, *args, **options):
+        orders = options['orders']
+        product_id = options['product_id']
+
+        if product_id:
```
Comment (clintonb) [inline ecommerce/core/management/commands/generate_bulk_orders.py]: Make this required and avoid the extra work.

File: ecommerce/core/management/commands/generate_bulk_orders.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                logger.exception(msg)
+                raise CommandError(msg)
+
+            if not product.is_seat_product:
+                msg = 'Product does not have seat.'
+                logger.exception(msg)
+                raise CommandError(msg)
+
+        else:
+            products = Product.objects.all()
```
Comment (clintonb) [inline ecommerce/core/management/commands/generate_bulk_orders.py]: This should go away but, in the future, just filter by product class if you only care about a single class of product.

File: ecommerce/core/management/commands/generate_bulk_orders.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+                    break
+
+            if not product:
+                msg = 'No product with seat exists.'
+                logger.exception(msg)
+                raise CommandError(msg)
+
+        user = User.objects.order_by('id').first()
+
+        for _ in range(orders):
```
Comment (clintonb) [inline ecommerce/core/management/commands/generate_bulk_orders.py]: Use double underscore, `__`. A single underscore, `_`, is typically reserved as an alias to `ugettext()`.

File: ecommerce/core/management/commands/generate_bulk_orders.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            basket = create_basket(user, product)
+            if not basket.id:
```
Comment (clintonb) [inline ecommerce/core/management/commands/generate_bulk_orders.py]: Given that you wrote `create_basket`, you should guarantee the basket has an ID. This should go away.

File: ecommerce/core/management/commands/generate_bulk_orders.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+                if product.is_seat_product:
+                    product = product
+                    break
+
+            if not product:
+                msg = 'No product with seat exists.'
+                logger.exception(msg)
+                raise CommandError(msg)
+
+        user = User.objects.order_by('id').first()
```
Comment (clintonb) [inline ecommerce/core/management/commands/generate_bulk_orders.py]: I would either create a new user, making this a great way to also populate the user table, or allow the user to pass in the username. I'm kinda leaning toward making a new one.

**Review by clintonb — CHANGES_REQUESTED** [review]

File: ecommerce/core/management/commands/generate_bulk_orders.py (hunk 4 above)
Comment (ahsan-ul-haq) [reply ecommerce/core/management/commands/generate_bulk_orders.py]: Are you referring to partner_sku?

Comment (clintonb) [reply ecommerce/core/management/commands/generate_bulk_orders.py]: Correct.

Comment (clintonb) [reply ecommerce/core/management/commands/generate_bulk_orders.py]: Also, note that you should probably use the partner from that SKU to get the `Site` that should be associated with the basket/order.

**Comment by ahsan-ul-haq** [comment]
@clintonb made the suggested changes.
FYI: @edx/learner-spartans

File: ecommerce/extensions/order/management/commands/create_fake_orders.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    help = 'Clean history data'
+
+    def add_arguments(self, parser):
+        parser.add_argument('--count',
+                            action='store',
+                            dest='orders',
+                            type=int,
+                            required=True,
+                            help='Number of orders to create.')
+        parser.add_argument('--partner_sku',
```
Comment (clintonb) [inline ecommerce/extensions/order/management/commands/create_fake_orders.py]: Nit: "sku" is sufficient.

File: ecommerce/extensions/order/management/commands/create_fake_orders.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+                            action='store',
+                            dest='partner_sku',
+                            type=str,
+                            required=True,
+                            help='Partner SKU for which orders are to be generated.')
```
Comment (clintonb) [inline ecommerce/extensions/order/management/commands/create_fake_orders.py]: SKU corresponding to the product for which orders will be created.

File: ecommerce/extensions/order/management/commands/create_fake_orders.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+            product = Product.objects.get(stockrecords__partner_sku=partner_sku)
+        except Product.DoesNotExist:
+            msg = 'No product for partner_sku {} exists.'.format(partner_sku)
+            logger.exception(msg)
+            raise CommandError(msg)
+
+        partner = StockRecord.objects.get(partner_sku=partner_sku).partner
+
+        try:
+            site = SiteConfiguration.objects.get(partner=partner).site
```
Comment (clintonb) [inline ecommerce/extensions/order/management/commands/create_fake_orders.py]: `site = partner.siteconfiguration.site`

File: ecommerce/extensions/order/management/commands/create_fake_orders.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+                            type=str,
+                            required=True,
+                            help='Partner SKU for which orders are to be generated.')
+
+    def handle(self, *args, **options):
+        orders = options['orders']
+        partner_sku = options['partner_sku']
+
+        try:
+            product = Product.objects.get(stockrecords__partner_sku=partner_sku)
```
Comment (clintonb) [inline ecommerce/extensions/order/management/commands/create_fake_orders.py]: Minor, but you can decrease the number of queries:

```python
stock_record = StockRecord.objects.get(partner_sku=partner_sku)
product = stock_record.product
partner = stock_record.partner
```

File: ecommerce/extensions/order/management/commands/create_fake_orders.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+            msg = 'No product for partner_sku {} exists.'.format(partner_sku)
+            logger.exception(msg)
+            raise CommandError(msg)
+
+        partner = StockRecord.objects.get(partner_sku=partner_sku).partner
+
+        try:
+            site = SiteConfiguration.objects.get(partner=partner).site
+        except SiteConfiguration.DoesNotExist:
+            msg = 'Site configuration for partner_sku {} does not exists.'.format(partner_sku)
```
Comment (clintonb) [inline ecommerce/extensions/order/management/commands/create_fake_orders.py]: Site configuration is linked to partners, not SKUs.

`No Site Configuration exists for partner {}!'.format(partner.id)`

File: ecommerce/extensions/order/management/commands/create_fake_orders.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+            logger.exception(msg)
+            raise CommandError(msg)
+
+        user = User.objects.create(username='fake_user', email='fake_user@example.com')
```
Comment (clintonb) [inline ecommerce/extensions/order/management/commands/create_fake_orders.py]: This will fail the second time this command is run. Use `user = UserFactory()`

File: ecommerce/extensions/order/management/commands/tests/test_create_fake_orders.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+
+LOGGER_NAME = 'ecommerce.extensions.order.management.commands.create_fake_orders'
+Order = get_model('order', 'Order')
+Basket = get_model('basket', 'Basket')
+
+
+class FakeOrdersTests(TestCase):
+
+    def test_product_not_exist(self):
```
Comment (clintonb) [inline ecommerce/extensions/order/management/commands/tests/test_create_fake_orders.py]: `test_product_does_not_exist`

File: ecommerce/extensions/order/management/commands/tests/test_create_fake_orders.py [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+                call_command('create_fake_orders', '--count=5', '--partner_sku=sku_11111')
+                log.check(
+                    (
+                        LOGGER_NAME,
+                        'EXCEPTION',
+                        'No product for partner_sku sku_11111 exists.'
+                    )
+                )
+
+    def test_site_config_not_exist(self):
```
Comment (clintonb) [inline ecommerce/extensions/order/management/commands/tests/test_create_fake_orders.py]: `test_site_configuration_does_not_exist`

**Review by clintonb — CHANGES_REQUESTED** [review]

File: ecommerce/extensions/order/management/commands/create_fake_orders.py (hunk 16 above)
Comment (ahsan-ul-haq) [reply ecommerce/extensions/order/management/commands/create_fake_orders.py]: May be we can use get_or_create to avoid failure I am bit uncomfortable with using factories here.

Comment (clintonb) [reply ecommerce/extensions/order/management/commands/create_fake_orders.py]: Write your own, non-test, factory then. The advantage of using the factory is that we get to create a large set of users. This is advantageous for times when we want to run tests on large user tables.

**Comment by ahsan-ul-haq** [comment]
@clintonb made the suggested changes

**Review by clintonb — APPROVED** [review]
